README.md
[![GitHub issues](https://img.shields.io/github/issues/Enngage/kontent-delivery-angular-sample)](https://github.com/Enngage/kontent-delivery-angular-sample/issues)
[![GitHub license](https://img.shields.io/github/license/Enngage/kontent-delivery-angular-sample)](https://github.com/Enngage/kontent-delivery-angular-sample/blob/master/LICENSE.md)
[![Discord](https://img.shields.io/discord/821885171984891914?label=Discord&logo=Discord&logoColor=white)](https://discord.gg/SKCxwPtevJ)
[![Build](https://github.com/Enngage/kontent-delivery-angular-sample/actions/workflows/main.yml/badge.svg)](https://github.com/Enngage/kontent-delivery-angular-sample/actions/workflows/main.yml)

# Kontent Angular Sample app

This is a sample Angular application showing the use of [@kentico/kontent-delivery](https://github.com/Kentico/kontent-delivery-sdk-js) SDK with a [Kontent Angular Http Service](https://github.com/Kentico/kontent-angular-http-service) for HTTP calls. 

## Live demo

See the app live at [https://enngage.github.io/kontent-delivery-angular-sample/](https://enngage.github.io/kontent-delivery-angular-sample/)

## Running the app locally

- Clone this repository
- Run `npm i `
- Run `npm start`

## Publishing demo app
Under root, generate demo app with

```
npm run build-demo-gh-pages
npx ngh --dir=dist
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.0](https://github.com/Enngage/KenticoKontentSampleAngularApp/compare/v0.0.1...v1.0.0) (2021-10-06)

### 0.0.1 (2021-10-06)


### Features

* implements new Angular sample app ([1d0de97](https://github.com/Enngage/KenticoKontentSampleAngularApp/commit/1d0de97ab1b9ba02e2ea7ce39e14c9e2e0a97487))
* installs standard version ([05ff681](https://github.com/Enngage/KenticoKontentSampleAngularApp/commit/05ff6815d68092351217eccb1cb14f3752c54306))


### Bug Fixes

* Resolved modular items not cleared in subsequent calls without reinitializing field map service  (https://github.com/Enngage/KenticoCloudDeliveryTypeScriptSDK/issues/46) ([2ec512f](https://github.com/Enngage/KenticoKontentSampleAngularApp/commit/2ec512f218d66c8fa2e3532386bd7b10cf94807b))


e2e/src/app.e2e-spec.ts
import { AppPage } from './app.po';
import { browser, logging } from 'protractor';

describe('workspace-project App', () => {
  let page: AppPage;

  beforeEach(() => {
    page = new AppPage();
  });

  it('should display welcome message', () => {
    page.navigateTo();
    expect(page.getTitleText()).toEqual('Welcome to kontent-poc-angular!');
  });

  afterEach(async () => {
    // Assert that there are no errors emitted from the browser
    const logs = await browser.manage().logs().get(logging.Type.BROWSER);
    expect(logs).not.toContain(jasmine.objectContaining({
      level: logging.Level.SEVERE,
    } as logging.Entry));
  });
});


src/app/news/news.component.spec.ts
import { TestBed, async } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AppComponent } from './app.component';

describe('AppComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      declarations: [
        AppComponent
      ],
    }).compileComponents();
  }));

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });

  it(`should have as title 'kontent-poc-angular'`, () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app.title).toEqual('kontent-poc-angular');
  });

  it('should render title in a h1 tag', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const compiled = fixture.debugElement.nativeElement;
    expect(compiled.querySelector('h1').textContent).toContain('Welcome to kontent-poc-angular!');
  });
});


src/karma.conf.js
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage-istanbul-reporter'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    coverageIstanbulReporter: {
      dir: require('path').join(__dirname, '../coverage/kontent-poc-angular'),
      reports: ['html', 'lcovonly', 'text-summary'],
      fixWebpackSourcePaths: true
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false,
    restartOnFileChange: true
  });
};


src/app/news/news.component.ts
import { ChangeDetectorRef, Component } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import {
  IDeliveryClient,
  createDeliveryClient,
  INetworkResponse,
  Responses,
  IGroupedNetworkResponse,
} from '@kentico/kontent-delivery';
import { AngularHttpService } from '@kentico/kontent-angular-http-service';
import { Observable, from } from 'rxjs';
import { observableHelper } from '../helpers/observable.helper';
import { getRequest } from '../helpers/http.helpers';
import { map } from 'rxjs/operators';

@Component({
  selector: 'app-news',
  templateUrl: './news.component.html',
  styleUrls: ['./news.component.css']
})
export class NewsComponent {
  newsData: any = {};
  displayName: string;
  isManager: boolean = false;
  id: string = "";

  deliveryClient: IDeliveryClient;

  itemsResponse?: INetworkResponse<Responses.IListContentItemsResponse, any>;
  itemResponse?: INetworkResponse<Responses.IViewContentItemResponse, any>;
  taxonomiesResponse?: INetworkResponse<Responses.IListTaxonomiesResponse, any>;
  taxonomyResponse?: INetworkResponse<Responses.IViewTaxonomyResponse, any>;
  typesResponse?: INetworkResponse<Responses.IListContentTypesResponse, any>;
  typeResponse?: INetworkResponse<Responses.IViewContentTypeResponse, any>;
  languagesResponse?: INetworkResponse<Responses.IListLanguagesResponse, any>;
  elementResponse?: INetworkResponse<Responses.IViewContentTypeElementResponse, any>;
  itemsFeedResponse?: IGroupedNetworkResponse<Responses.IListItemsFeedAllResponse>;

  constructor(private router : Router, private http: HttpClient, private httpClient: HttpClient, private cdr: ChangeDetectorRef) {
    this.deliveryClient = createDeliveryClient({
      projectId: '27fea92e-2633-0035-9c21-b601899437c5',
      httpService: new AngularHttpService(httpClient),
    });
    this.displayName = "";
   }

  async ngOnInit() {    
     this.id = this.router.url.split("/")[this.router.url.split("/").length-1];
      this.getCollection();
  }


  private getCollection(): void {
    this.zipAndExecute([
      from(this.deliveryClient.items()
      .equalsFilter('system.codename', this.id.toLowerCase())
      .toPromise()).pipe(
        map((response) => {
          this.itemsResponse = response;
          console.log('this.itemsResponse = '+ JSON.stringify(this.itemsResponse));
          response.data.items.forEach((element) => { 
          this.newsData['title'] = element.elements.title.value;
          this.newsData['description'] = element.elements.description.value;
          this.newsData['image'] = element.elements.image.value[0].url;
          });
          this.cdr.markForCheck();
        })
      ),
      ]);
  }

  private zipAndExecute(observables: Observable<void>[]): void {
    observableHelper
      .zipObservables(observables)
      .pipe(
        map(() => {
          this.cdr.markForCheck();
        })
      )
      .subscribe();
  }

}


src/app/helpers/http.helpers.ts
const handleErrorResponse = async (response: Response) => {
    const contentType = response.headers.get('content-type');

    if (contentType && contentType.indexOf('application/json') !== -1) {
        const errorMessage = await readJSON(response);
        throw new Error(JSON.stringify(errorMessage));
    } else {
        const errorMessage = await readText(response);
        throw new Error(errorMessage);
    }
}

const readJSON = (response: Response) => {
    return response.json();
}

const readText = (response: Response) => {
    return response.text();
}

const handleResponse = (response: Response) => {
    if (!response.ok) {
        return handleErrorResponse(response);
}

const contentType = response.headers.get('content-type');
    if (contentType && contentType.indexOf('application/json') !== -1) {
        return readJSON(response);
    } else {
        return readText(response);
    }
}

export const getRequest = <T>(url: string, user: any): Promise<T> => {
    if(null != user){
        let headers = new Headers();
        headers.append("SMUSER", user);
        var requestOptions = {
            method: 'GET',
            headers: headers
        };
        return fetch(url, requestOptions).then(handleResponse);
    } else {
        return fetch(url).then(handleResponse);
    }
}

src/app/tile/tile.component.ts
import { ChangeDetectorRef, Component } from '@angular/core';
import {
  IDeliveryClient,
  createDeliveryClient,
  INetworkResponse,
  Responses,
} from '@kentico/kontent-delivery';
import { Observable, from } from 'rxjs';
import { observableHelper } from '../helpers/observable.helper';
import { getRequest } from '../helpers/http.helpers';
import { map } from 'rxjs/operators';

@Component({
  selector: 'app-tile',
  templateUrl: './tile.component.html',
  styleUrls: ['./tile.component.css']
})
export class TileComponent {
  title = '';
  carouselData: any = [];
  displayName: string = "";
  isManager: boolean = false;
  showUserName: boolean = false;
  userId: string = "";

  deliveryClient: IDeliveryClient;
  itemsResponse?: INetworkResponse<Responses.IListContentItemsResponse, any>;
  itemResponse?: INetworkResponse<Responses.IViewContentItemResponse, any>;

  constructor(private cdr: ChangeDetectorRef) {
    this.deliveryClient = createDeliveryClient({
      projectId: '27fea92e-2633-0035-9c21-b601899437c5'
    });
   }
  
   bindUserId(e: any) {
    this.userId = e.target.value;
   }

  async getUser(e: any){
    let userData: any = await this.getUserID(this.userId);
    this.isManager = userData.MANAGER? true : false;
    this.displayName = userData.DISPLAY_NAME;
    this.showUserName = true;
    this.getCollection();
  }

  async getUserID(id: any){
    return getRequest("/pulseprofile/api/v1/users/"+id, null);
  }

  private getCollection(): void {
    let roles = ['associate', 'consultant'];
    if(this.isManager){
      roles.push('manager');
    }
    this.zipAndExecute([
      from(this.deliveryClient.items()
      .equalsFilter('system.codename', 'news_tile')
      .anyFilter('elements.roles', roles)
      // .orderParameter('system.last_modified', 'desc')
      .toPromise()).pipe(
        map((response: any) => {
          this.itemsResponse = response;
          console.log('this.itemsResponse = '+ JSON.stringify(this.itemsResponse));
          response.data.items[0].elements.news_items.linkedItems.forEach((element: any) => {
            let roles: string = '';
            element.elements.roles.value.forEach((role: any) => {
              if(role.codename.indexOf('manager') == -1){
                roles = roles+role.codename+',';
              }
            });
            if(this.isManager || roles != ""){
              let carouselItem: any = {};
              carouselItem['title'] = element.elements.title.value;
              carouselItem['image'] = element.elements.image.value[0].url;
              carouselItem['link'] = '/news/'+element.system.codename.toUpperCase();
              this.carouselData.push(carouselItem);
            }
          });
          this.cdr.markForCheck();
        })
      ),
      ]);
  }

  private zipAndExecute(observables: Observable<void>[]): void {
    observableHelper
      .zipObservables(observables)
      .pipe(
        map(() => {
          this.cdr.markForCheck();
        })
      )
      .subscribe();
  }

}
